### PR TITLE
Move JNDI `DataSource` to separate module

### LINKED ISSUE #1914: Split JNDI support out of `log4j-jdbc`
Currently the `log4j-jdbc` appender has an **optional** dependency on `log4j-jndi`.

In order to remove this optional dependency, I propose to:

- create a new module `log4j-jdbc-jndi` and depends on `log4j-jdbc` and `log4j-jndi` and contains the [`DataSourceConnectionSource`](https://github.com/apache/logging-log4j2/blob/main/log4j-jdbc/src/main/java/org/apache/logging/log4j/jdbc/appender/DataSourceConnectionSource.java)  and inlined code from `JndiUtil`,
- remove the dependency on `log4j-jndi` from `log4j-jdbc`.

From a user perspective:

- most users will require just the `log4j-jdbc-dbcp2` module,
- some Java EE/Jakarta EE users will require the `log4j-jdbc-jndi` module,
- the `log4j-jdbc` module could be considered as an internal dependency.

### PR DESCRIPTION
**Pull request by ppkarwasz**
We move the [`DataSource` connection source](https://logging.staged.apache.org/log4j/3.x/manual/appenders/database.html#DataSourceConnectionSource) to a new module `log4j-jdbc-jndi`.

The main rationale for this change is to remove the **optional** dependency on `log4j-jndi` from `log4j-jdbc`.
This change forces users to opt-in for the JNDI `DataSource` by adding `log4j-jdbc-jndi` to their classpath.

After this change, the `log4j-jdbc` artifact contains:

* the `JDBC` Appender,
* a connection source based on `DriverManager`,
* a connection source based on a static factory method.

Closes #1914

### CODE REVIEW

File: src/changelog/.3.x.x/1914_split_JDBC_JNDI_connection_source.xml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<entry xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+       xmlns="https://logging.apache.org/xml/ns"
+       xsi:schemaLocation="https://logging.apache.org/xml/ns https://logging.apache.org/xml/ns/log4j-changelog-0.xsd"
+       type="added">
+  <issue id="1914" link="https://github.com/apache/logging-log4j2/issues/1914"/>
+  <description format="asciidoc">Split off JDBC `DataSource` connection source into a new `log4j-jdbc-jndi` module.</description>
```
Comment (vy) [inline src/changelog/.3.x.x/1914_split_JDBC_JNDI_connection_source.xml:7]: Umm... I think this is not correct. Suggestion:

```suggestion
  <description format="asciidoc">Split off JNDI support of `log4j-jdbc` to a new `log4j-jdbc-jndi` module</description>
```

File: log4j-jdbc/pom.xml [hunk 2]
```diff
@@ -30,89 +30,99 @@
...
-      <scope>test</scope>
+      <artifactId>log4j-plugins</artifactId>
     </dependency>
+
     <dependency>
       <groupId>org.apache.logging.log4j</groupId>
       <artifactId>log4j-core-test</artifactId>
       <scope>test</scope>
+      <!--
+        ~ Exclude some unused dependencies of `log4j-core-test`.
```
Comment (vy) [inline log4j-jdbc/pom.xml:55]: Why do we need this?

**Review by vy — COMMENTED** [review]
Overall, LGTM.

**Question:** Is the DBCP detail of our `log4j-jdbc*` modules well-hidden? If so, I suggest

1. Removing all of its public mentions
2. Switching to HikariCP<sup>1</sup> instead

<sup>1</sup> HikariCP is, both performance- and feature-wise, a clear winner against any other alternative in the market as of today. It is the default JDBC connection pool in both Spring Boot 2 and [Spring Boot 3](https://docs.spring.io/spring-boot/reference/data/sql.html#data.sql.datasource.connection-pool).

Of course I am not suggesting this for this PR. But I feel like this is an okay place to discuss it.

File: log4j-jdbc/pom.xml (hunk 2 above)
Comment (ppkarwasz) [reply log4j-jdbc/pom.xml:55]: Right, we don't. The `log4j-core-test` artifact has a lot of dependencies that are added to the test classpath of each of our components. We probably should remove those that are not necessary.

Fixed in https://github.com/apache/logging-log4j2/pull/2915/commits/c9cc897bd2142eded6dfa475446cbe805064e1a8

Comment (vy) [reply log4j-jdbc/pom.xml:55]: Removed from the wrong module.

Comment (vy) [reply log4j-jdbc/pom.xml]: Fixed in af3b21b067a2546f089716d16a2a57ee0292d517.

**Review by vy — APPROVED** [review]

**Comment by ppkarwasz** [comment]
> **Question:** Is the DBCP detail of our `log4j-jdbc*` modules well-hidden? If so, I suggest
> 
>     1. Removing all of its public mentions
> 
>     2. Switching to HikariCP instead

DBCP has a nice feature, that justifies its pool of _aficionados_: you can register a pool with the JDBC driver and retrieve it later using a `jdbc:apache:commons` URI.

Note that until #1916 is fixed, all pools perform the same: only a single connection is used at each time.
The main problem is the synchronization of `AbstractDatabaseManager`. Removing it might technically not be a breaking change, but it is a very **disruptive** one.

**Pull request merged**